app/domain/authentication/authn_k8s/application_identity.rb

```diff
@@ -0,0 +1,181 @@
+    # valid and will fail the validation here. However, a application identity
+    # `some-namespace/service-account/some-value` is a valid application identity
+    # and will pass the validation here. If the host is actually running from
+    # a pod with service account `some-other-value` then it will fail the
+    # validation of ValidateApplicationIdentity
+    class ApplicationIdentity
+
+      def initialize(host_id:, host_annotations:, service_id:)
+        @host_id          = host_id
+        @host_annotations = host_annotations
+        @service_id       = service_id
+
+        validate
+      end
+
+      def namespace
+        @namespace ||= application_identity_in_annotations? ? constraint_from_annotation("namespace") : host_id_suffix[-3]
+      end
+
+      def constraints
+        @constraints ||= {
+          pod:               constraint_value("pod"),
+          service_account:   constraint_value("service_account"),
+          deployment:        constraint_value("deployment"),
+          deployment_config: constraint_value("deployment_config"),
+          stateful_set:      constraint_value("stateful_set")
+        }
+      end
+
+      def container_name
+        annotation_name = "authentication-container-name"
+        annotation_value("authn-k8s/#{@service_id}/#{annotation_name}") ||
+          annotation_value("authn-k8s/#{annotation_name}") ||
+          annotation_value("kubernetes/#{annotation_name}") ||
+          "authenticator"
+      end
+
+      def namespace_scoped?
+        constraints.compact.empty?
+      end
+
+      private
+
+      # Validates that the application identity is defined correctly
+      def validate
+        validate_permitted_scope
+        validate_namespace_constraint_exists
+        validate_constraint_combinations
+      end
+
+      # If the application identity is defined in:
+      #   - annotations: validates that all the constraints are
+      #                  valid (e.g there is no "authn-k8s/blah" annotation)
+      #   - host id: validates that the host-id has 3 parts and that the given
+      #              constraint is valid (e.g the host id is not
+      #              "namesapce/blah/some-value")
+      def validate_permitted_scope
+        application_identity_in_annotations? ? validate_permitted_annotations : validate_host_id
+      end
+
+      # This method assumes that 'validate_permitted_scope' has already run.
+      # It uses the value of 'host_id_suffix' after it is validated.
+      def validate_namespace_constraint_exists
+        return if namespace
+        raise Err::MissingNamespaceConstraint
+      end
+
+      # Validates that the application identity doesn't include logical constraint
+      # combinations (e.g deployment & deploymentConfig)
+      def validate_constraint_combinations
+        return if permitted_constraint_combinations?
+        raise Err::IllegalConstraintCombinations
+      end
+
+      def permitted_constraint_combinations?
```

Comment:
Why not raise the `Err::IllegalConstraintCombinations` message in this method rather than adding a second method (above) to perform the check?

Reply:
i left only the method above and moved the body into it:
```
      # Validates that the application identity doesn't include logical constraint
      # combinations (e.g deployment & deploymentConfig)
      def validate_constraint_combinations
        # validate that we have only one constraint from the following
        controllers = %i(deployment deployment_config stateful_set)

        unless (constraints.compact.keys & controllers).length <= 1
          raise Err::IllegalConstraintCombinations
        end
      end
```